src/stacked_borrows/stack.rs

```diff
@@ -0,0 +1,363 @@
+            if cache_idx == 1 {
+                self.cache.tags.swap(0, 1);
+                self.cache.idx.swap(0, 1);
+            } else if cache_idx > 1 {
+                self.cache.add(stack_idx, tag);
+            }
+            Some(stack_idx)
+        } else {
+            // Tag is in the cache, but it doesn't grant the required permission
+            None
+        }
+    }
+
+    pub fn insert(&mut self, new_idx: usize, new: Item) {
+        self.borrows.insert(new_idx, new);
+
+        #[cfg(feature = "stack-cache")]
+        self.insert_cache(new_idx, new);
+    }
+
+    #[cfg(feature = "stack-cache")]
+    fn insert_cache(&mut self, new_idx: usize, new: Item) {
+        // Adjust the possibly-unique range if an insert occurs before or within it
+        if self.unique_range.start >= new_idx {
+            self.unique_range.start += 1;
+        }
+        if self.unique_range.end >= new_idx {
+            self.unique_range.end += 1;
+        }
+        if new.perm == Permission::Unique {
+            // Make sure the possibly-unique range contains the new borrow
+            self.unique_range.start = self.unique_range.start.min(new_idx);
+            self.unique_range.end = self.unique_range.end.max(new_idx + 1);
+        }
+
+        // The above insert changes the meaning of every index in the cache >= new_idx, so now
+        // we need to find every one of those indexes and increment it.
+        // But if the insert is at the end (equivalent to a push), we can skip this step because
+        // it didn't change the position of any other tags.
+        if new_idx != self.borrows.len() - 1 {
+            for idx in &mut self.cache.idx {
+                if *idx >= new_idx {
+                    *idx += 1;
+                }
+            }
+        }
+
+        // This primes the cache for the next access, which is almost always the just-added tag.
+        self.cache.add(new_idx, new.tag);
+
+        #[cfg(feature = "expensive-debug-assertions")]
+        self.verify_cache_consistency();
+    }
+
+    /// Construct a new `Stack` using the passed `Item` as the base tag.
+    pub fn new(item: Item) -> Self {
+        Stack {
+            borrows: vec![item],
+            unknown_bottom: None,
+            #[cfg(feature = "stack-cache")]
+            cache: StackCache { idx: [0; CACHE_LEN], tags: [item.tag; CACHE_LEN] },
+            #[cfg(feature = "stack-cache")]
+            unique_range: if item.perm == Permission::Unique { 0..1 } else { 0..0 },
+        }
+    }
+
+    pub fn get(&self, idx: usize) -> Option<Item> {
+        self.borrows.get(idx).cloned()
+    }
+
+    #[allow(clippy::len_without_is_empty)] // Stacks are never empty
+    pub fn len(&self) -> usize {
+        self.borrows.len()
+    }
+
+    pub fn unknown_bottom(&self) -> Option<SbTag> {
+        self.unknown_bottom
+    }
+
+    pub fn set_unknown_bottom(&mut self, tag: SbTag) {
+        self.borrows.clear();
```

Comment:
No cache to clear here?

Reply:
The cache can't be cleared, because it doesn't have a dynamic length so I've tried to explain in a commit why not clearing it is okay.

I get that this is a bit of a complicated requirement to uphold, but on the other hand that is why we have `verify_cache_consistency`. Giving the cache a length would be similar to making it an `arrayvec` and I'm not sure that's a win.